### ISSUE TITLE
Methods for setting options are inconsistent

### ISSUE THREAD

**Issue by prattmic** [restic/restic#1233]
The different way to set options varies widely from option to option.

* Some can be set via a flag or environment variable. e.g., `--repo`/`RESTIC_REPOSITORY`

* Some can be set only via a flag. e.g., `--json`, `--hostname`

* Some backend options can only be set via `--options`. e.g., `--options connection=10`

* Some backend options can only be set via environment variables. e.g., `GOOGLE_PROJECT_ID`

IMO, the options should be consistent. That is, they should all be settable via flags or environment variables.

This is something I can try to clean up and send a PR for when I get a chance.

**Comment by pvgoran** [comment]
Just my humble opinion: personally I don't see any problem with this diversity, it seems quite natural, except for the point 4 (if a backend option is settable via an environment variable, it should be settable from the command line). In particular, setting backend options via `--options` looks like a better approach in comparison with setting them via first-class command line options.

That said, the environment variable for the `--hostname` command line option looks like a welcome addition.

**Comment by prattmic** [comment]
Agreed, I should have been more clear. I don't think that backend options need to be raised up to full dedicated CLI options, they are fine behind --options. It is the environment variable behavior that should be consistent.

**Comment by fd0** [comment]
Thanks for raising this issue. I admit that there's a bit of inconsistency, so I'll try to give you some background: There are two places where environment variables are used:
 * For global options that need to be set for each command: `RESTIC_REPOSITORY` and `RESTIC_PASSWORD`, export them once and then run multiple restic commands
 * For the backend configuration: This was originally added because in the AWS S3 world it's common to have the access credentials in the environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`, so we just used that. And it seems quite common for other tools too. I agree with you that it's a good idea to have a way to use `--option` to set this, e.g. with `--option s3.access_key_id=foo` or so. The extended options (see below) were added quite recently, and not present during initial implementation of the backends.

Then there are the extended options, `--option foo=bar`, which have been added recently. This is similar to what e.g. `ssh` or `git` do, and I expect that we will add many more options in the future that either are not important enough to be promoted to top-level options or are within a specific namespace (e.g. `s3.connections=`).

However, I disagree that every option should be set via an environment variable. For example, what's the use case of being able to specify the hostname for a new backup via an environment variable instead of the option `--hostname`? It does not fit into either of the categories mentioned above: It's not relevant to more than one command, and it's not a backend configuration option either.

Is there a use case that I don't see right now?

**Comment by prattmic** [comment]
My use case is for a package I'm writing to control restic. It is cleanest for it to simply set all options the command may need, like hostname, via an environment variable rather than tracking that some must be flags on specific commands.

Doing so is definitely not the end of the world, it just feels inconsistent.

**Comment by ghjm** [comment]
I would like to be able to set b2.connections in the environment - or at least somewhere permanent - so that I don't have to remember to specify it _every single time_ I run a restic command.

**Comment by calebcase** [comment]
I found this confusing as well. I thought I could set the path to the rclone program through an environment variable `RESTIC_RCLONE_PROGRAM`, but it seems to only be possible through `--option rclone.program=/some/path`.

**Comment by thmo** [comment]
Here's a use case for specifying the hostname via an environment variable: using `restic` via `docker-compose`. It would allow setting the desired hostname already in the `docker-compose.yml`, for convenience, overriding the random hostname generated by docker.
